crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4142,23 +4159,46 @@ mod tests {
         assert_public_ty(&db, "src/a.py", "d", "Literal[-1]");
         assert_public_ty(&db, "src/a.py", "e", "Literal[2]");
 
+        // More division cases
         db.write_dedented(
             "/src/b.py",
             "
             a = 1 / 2
             b = 1 // 2
             c = 4 / 2
             d = 4 // 2
-            e = 4 // 0
             ",
         )?;
 
         assert_public_ty(&db, "/src/b.py", "a", "float");
         assert_public_ty(&db, "/src/b.py", "b", "Literal[0]");
         assert_public_ty(&db, "/src/b.py", "c", "float");
         assert_public_ty(&db, "/src/b.py", "d", "Literal[2]");
-        // TODO: division by zero should be an error
-        assert_public_ty(&db, "/src/b.py", "e", "int");
+
+        // Division by zero
+        db.write_dedented(
+            "/src/c.py",
+            "
+            a = 1 / 0
+            b = 2 // 0
+            c = 3 % 0
+            ",
+        )?;
+
+        assert_public_ty(&db, "/src/c.py", "a", "Unknown");
+        assert_public_ty(&db, "/src/c.py", "b", "Unknown");
+        assert_public_ty(&db, "/src/c.py", "c", "Unknown");
+
+        assert_file_diagnostics(
+            &db,
+            "src/c.py",
+            &[
+                "Cannot divide type 'Literal[1]' by zero.",
+                "Cannot divide type 'Literal[2]' by zero.",
+                "Cannot divide type 'Literal[3]' by zero.",
```

Comment:
We could want a special message for modulo.

Reply:
Modulo requires division to calculate, so I think the error message is accurate and adequate as is.

Reply:
Agree, but not all beginners may realize that! Seems fine unless we receive significant feedback though.

Reply:
Yeah I agree with @zanieb here -- the current message is entirely accurate, but I think even your average Python user doesn't necessarily think about modulo operations in terms of their desugared sub-operations, so I think encountering this error message here might be pretty surprising to a lot of our users. It's definitely not the biggest issue in the world, but I think a better error message here would be really worthwhile (and I actually think it's probably better to do it now before we forget about it)